oci/default.nix

```diff
@@ -0,0 +1,556 @@
+        rm $out/db/reserved
+      '';
+    in
+    {
+      "nix/var/nix" = {
+        type = "directory";
+        sources = [{
+          path = store;
+        }];
+      };
+    };
+
+  /* Wraps a derivation such that including it in a layer only includes its
+     dependencies
+  */
+  makeDependencyOnlyWrapper = drv:
+    let
+      referencesFile = writeReferencesToFile drv;
+    in
+    runCommand "${drv.name}-dependencies" {
+      passthru.excludeFromLayer = true;
+    } ''
+      grep -v "${drv}" "${referencesFile}" > $out
+    '';
+
+  /* Creates a layer from entries without its dependencies
+
+     This is mainly for internal use. See makeLayer for the structure of
+     entries.
+  */
+  makeBareLayerFromEntries = {
+    name,
+    entries,
+    umask
+  }: runCommand name {
+    entriesJson = builtins.toJSON entries;
+    passAsFile = [ "entriesJson" ];
+    passthru.excludeFromLayer = true;
+  } ''
+    mkdir -p $out
+    ${python3}/bin/python ${./build-layer.py} $out ${umask} "$SOURCE_DATE_EPOCH" < $entriesJsonPath
+    ${python3}/bin/python ${./hash-layer.py} $out
+  '';
+
+  /* Creates a layer from an optional base tar file and include/excludes
+
+     This is mainly for internal use. Prefer using makeLayer instead.
+  */
+  makeLayer' = {
+    name,
+    baseTar ? null,
+    includes ? [],
+    excludes ? [],
+    passthru ? {}
+  }:
+    let
+      includesFile = writeText "layer-includes" (builtins.concatStringsSep "\n" includes);
+      excludesFile = writeText "layer-excludes" (builtins.concatStringsSep "\n" excludes);
+      directExcludes = builtins.filter (x: x.excludeFromLayer or false) includes;
+    in
+    runCommand name {
+      layerIncludes = writeReferencesToFile includesFile;
+      layerExcludes = writeReferencesToFile excludesFile;
+      directExcludes = directExcludes ++ [ includesFile ];
+      inherit baseTar;
+
+      passthru = passthru // {
+        inherit includes;
+
+        # When building things layer by layer, we typically want to exclude all
+        # previous layer's included dependencies. Specifying just
+        # lastLayer.propagatedDependencies is easier instead of having to
+        # combine all previous layer's includes.
+        propagatedDependencies = includes ++ excludes;
+      };
+    } ''
+      cp --no-preserve=mode $layerExcludes excludePaths
+
+      # don't copy things we only use for deriving references
+      for exclude in $directExcludes; do
+        echo "$exclude" >> excludePaths
+      done
+
+      comm <(sort excludePaths) <(sort $layerIncludes) -1 -3 > pathsToCopy
+
+      mkdir -p $out
+      touch tarPaths
+      if [ -n "$baseTar" ] && [ -f "$baseTar" ]; then
+        cp --no-preserve=mode "$baseTar" "$out/layer.tar"
+        tar tf "$baseTar" > tarPaths
+      fi
+
+      tar_with_opts() {
+        tar rf $out/layer.tar \
+          --sort=name \
+          --mtime="@$SOURCE_DATE_EPOCH" \
+          --owner=0:0 \
+          --group=0:0 \
```

Comment:
This sets both user and uid to 0. I suppose it doesn't matter too much for reproducibility since 0 is always root (citation needed), but just to be consistent with how the python script does it, I'm making it match too.